## EDA And Feature Engineering For Flight Price Prediction Dataset)

### FEATURES

#### The various features of the cleaned dataset are explained below:

1) Airline: The name of the airline company is stored in the airline column & is a categorical feature having 6 diff airlines.
 
2) Flight: Flight stores information regarding the plane's flight code. It is a categorical feature. 
 
3) Source City: City from which the flight takes off. It is a categorical feature having 6 unique cities. 
 
4) Departure Time:  It stores information about the departure time and have 6 unique time label.
 
5) Stops: A categorical feature with 5 distinct values storing the number of stops between the source and destination cities. 

6) Arrival Time: It has six distinct time labels and keeps information about the arrival time.

7) Destination City: City where the flight will land. It is a categorical feature having 6 unique cities. 

8) Class: A categorical feature that contains information on seat class; it has two distinct values: Business and Economy. 

9) Duration: A continuous feature that displays the overall amount of time it takes to travel between cities in hours. 

10) Days Left: This is a derived characteristic that is calculated by subtracting the trip date by the booking date.

11) Price: Target variable stores information of the ticket price. 

In [3]:
# Importing several libraries, including pandas, numpy, matplotlib, and seaborn, which are commonly used in data analysis and visualization.

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
df=pd.read_excel('flight_price.xlsx') #Reads an Excel file called 'flight_price.xlsx' using the pd.read_excel() function.

## DATA OVERVIEW

In [2]:
## To get the top 5 rows just to know how dataset looks like.
df.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

In [4]:
# To know the shape of the dataset
df.shape

(10683, 11)

In [6]:
## To get the basics info about data

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


In [7]:
## To know about the unique value in each column of a dataset.
df.nunique()

Airline              12
Date_of_Journey      44
Source                5
Destination           6
Route               128
Dep_Time            222
Arrival_Time       1343
Duration            368
Total_Stops           5
Additional_Info      10
Price              1870
dtype: int64

In [8]:
# to know the count of different values
df.Destination.unique(),df.Source.unique()

(array(['New Delhi', 'Banglore', 'Cochin', 'Kolkata', 'Delhi', 'Hyderabad'],
       dtype=object),
 array(['Banglore', 'Kolkata', 'Delhi', 'Chennai', 'Mumbai'], dtype=object))

In [10]:
df.Airline.unique()

array(['IndiGo', 'Air India', 'Jet Airways', 'SpiceJet',
       'Multiple carriers', 'GoAir', 'Vistara', 'Air Asia',
       'Vistara Premium economy', 'Jet Airways Business',
       'Multiple carriers Premium economy', 'Trujet'], dtype=object)

In [11]:
df.Total_Stops.unique()

array(['non-stop', '2 stops', '1 stop', '3 stops', nan, '4 stops'],
      dtype=object)

In [12]:
df['Arrival_Time'].unique()

array(['01:10 22 Mar', '13:15', '04:25 10 Jun', ..., '06:50 10 Mar',
       '00:05 19 Mar', '21:20 13 Mar'], dtype=object)

In [13]:
df['Dep_Time'].unique()

array(['22:20', '05:50', '09:25', '18:05', '16:50', '09:00', '18:55',
       '08:00', '08:55', '11:25', '09:45', '20:20', '11:40', '21:10',
       '17:15', '16:40', '08:45', '14:00', '20:15', '16:00', '14:10',
       '22:00', '04:00', '21:25', '21:50', '07:00', '07:05', '09:50',
       '14:35', '10:35', '15:05', '14:15', '06:45', '20:55', '11:10',
       '05:45', '19:00', '23:05', '11:00', '09:35', '21:15', '23:55',
       '19:45', '08:50', '15:40', '06:05', '15:00', '13:55', '05:55',
       '13:20', '05:05', '06:25', '17:30', '08:20', '19:55', '06:30',
       '14:05', '02:00', '09:40', '08:25', '20:25', '13:15', '02:15',
       '16:55', '20:45', '05:15', '19:50', '20:00', '06:10', '19:30',
       '04:45', '12:55', '18:15', '17:20', '15:25', '23:00', '12:00',
       '14:45', '11:50', '11:30', '14:40', '19:10', '06:00', '23:30',
       '07:35', '13:05', '12:30', '15:10', '12:50', '18:25', '16:30',
       '00:40', '06:50', '13:00', '19:15', '01:30', '17:00', '10:00',
       '19:35', '15:

In [14]:
df['Duration'].value_counts(sort=True)

2h 50m     550
1h 30m     386
2h 45m     337
2h 55m     337
2h 35m     329
          ... 
31h 30m      1
30h 25m      1
42h 5m       1
4h 10m       1
47h 40m      1
Name: Duration, Length: 368, dtype: int64

In [15]:
df['Additional_Info'].value_counts(normalize=True)*100

No info                         78.114762
In-flight meal not included     18.552841
No check-in baggage included     2.995413
1 Long layover                   0.177853
Change airports                  0.065525
Business class                   0.037443
No Info                          0.028082
1 Short layover                  0.009361
Red-eye flight                   0.009361
2 Long layover                   0.009361
Name: Additional_Info, dtype: float64

Conclusion-

1)There a in total 10683 rows and 11 columns in the dataset.
2)There are 12 unique airlines.
3)There are 5 unique source cities and 6 unique Destination cities.
4)Total of 128 different routes.

Cleaning-

1)In route and total stops column there is one null value.
2)In destination cities New Delhi is denoted as both delhi and New delhi 
3)In duration column there is a value showing 5m to be looked upon.
4)Arrival time column needs to be cleaned as in some columns there is time along with dates.
5)Dropping of additional info column as 80 % values are no info.


## DATA CLEANING

In [16]:
df1=df.copy()

In [17]:
# To show only the rows that have at least one missing value (i.e., NaN) in any column
df1[df1.isnull().any(axis=1)]

Airline Date_of_Journey Source Destination Route Dep_Time  \
9039  Air India       6/05/2019  Delhi      Cochin   NaN    09:45   

      Arrival_Time Duration Total_Stops Additional_Info  Price  
9039  09:25 07 May  23h 40m         NaN         No info   7480

In [18]:
# To count the number of times each unique value appears in the 'Total_Stops' column to fill the null value for above row.

df1[(df1['Airline']=='Air India') & (df1['Source']=="Delhi") & (df1['Destination']=="Cochin")]['Total_Stops'].value_counts()

2 stops     373
1 stop      281
non-stop     75
3 stops      17
Name: Total_Stops, dtype: int64

In [19]:
#Imputing missing values in flight price data based on the mode of the number of stops.
 
df1.loc[9039,'Total_Stops']='2 stops'
df1.loc[9039]

Airline               Air India
Date_of_Journey       6/05/2019
Source                    Delhi
Destination              Cochin
Route                       NaN
Dep_Time                  09:45
Arrival_Time       09:25 07 May
Duration                23h 40m
Total_Stops             2 stops
Additional_Info         No info
Price                      7480
Name: 9039, dtype: object

In [20]:
df1[(df1['Airline']=='Air India') & (df1['Source']=="Delhi") & (df1['Destination']=="Cochin") & (df1['Total_Stops']=="2 stops") ]['Route'].value_counts()

DEL → AMD → BOM → COK    64
DEL → GOI → BOM → COK    58
DEL → HYD → BOM → COK    56
DEL → HYD → MAA → COK    47
DEL → CCU → BOM → COK    43
DEL → JAI → BOM → COK    33
DEL → BHO → BOM → COK    21
DEL → LKO → BOM → COK    16
DEL → JDH → BOM → COK    14
DEL → UDR → BOM → COK    11
DEL → NAG → BOM → COK    10
Name: Route, dtype: int64

In [21]:
df1.loc[9039,'Route']='DEL → AMD → BOM → COK'
df1.loc[9039]

Airline                        Air India
Date_of_Journey                6/05/2019
Source                             Delhi
Destination                       Cochin
Route              DEL → AMD → BOM → COK
Dep_Time                           09:45
Arrival_Time                09:25 07 May
Duration                         23h 40m
Total_Stops                      2 stops
Additional_Info                  No info
Price                               7480
Name: 9039, dtype: object

In [22]:
df1.isnull().sum()

Airline            0
Date_of_Journey    0
Source             0
Destination        0
Route              0
Dep_Time           0
Arrival_Time       0
Duration           0
Total_Stops        0
Additional_Info    0
Price              0
dtype: int64

In [23]:
# Replacing 'Delhi' with 'New Delhi' to avoid redundancy in the data.

df1['Source'] = df1['Source'].replace('Delhi', 'New Delhi')
df1[(df1['Source']=="Delhi")]

Empty DataFrame
Columns: [Airline, Date_of_Journey, Source, Destination, Route, Dep_Time, Arrival_Time, Duration, Total_Stops, Additional_Info, Price]
Index: []

In [24]:
df1['Destination'] = df1['Destination'].replace('Delhi', 'New Delhi')
df1[(df1['Destination']=="Delhi")]

Empty DataFrame
Columns: [Airline, Date_of_Journey, Source, Destination, Route, Dep_Time, Arrival_Time, Duration, Total_Stops, Additional_Info, Price]
Index: []

In [25]:
df1.Destination.unique(),df1.Source.unique()

(array(['New Delhi', 'Banglore', 'Cochin', 'Kolkata', 'Hyderabad'],
       dtype=object),
 array(['Banglore', 'Kolkata', 'New Delhi', 'Chennai', 'Mumbai'],
       dtype=object))

In [26]:
df1[df1['Duration']=='5m']
df1.loc[6474,'Duration']='24h 5m'
df1.loc[6474]

Airline                        Air India
Date_of_Journey                6/03/2019
Source                            Mumbai
Destination                    Hyderabad
Route              BOM → GOI → PNQ → HYD
Dep_Time                           16:50
Arrival_Time                       16:55
Duration                          24h 5m
Total_Stops                      2 stops
Additional_Info                  No info
Price                              17327
Name: 6474, dtype: object

In [27]:
df1['Arrival_Time']=df1['Arrival_Time'].str.split(' ').str[0]
df1['Arrival_Time']

0        01:10
1        13:15
2        04:25
3        23:30
4        21:35
         ...  
10678    22:25
10679    23:20
10680    11:20
10681    14:10
10682    19:15
Name: Arrival_Time, Length: 10683, dtype: object

In [28]:
df1.drop('Additional_Info',axis=1,inplace=True)
df1

Airline Date_of_Journey     Source Destination  \
0           IndiGo      24/03/2019   Banglore   New Delhi   
1        Air India       1/05/2019    Kolkata    Banglore   
2      Jet Airways       9/06/2019  New Delhi      Cochin   
3           IndiGo      12/05/2019    Kolkata    Banglore   
4           IndiGo      01/03/2019   Banglore   New Delhi   
...            ...             ...        ...         ...   
10678     Air Asia       9/04/2019    Kolkata    Banglore   
10679    Air India      27/04/2019    Kolkata    Banglore   
10680  Jet Airways      27/04/2019   Banglore   New Delhi   
10681      Vistara      01/03/2019   Banglore   New Delhi   
10682    Air India       9/05/2019  New Delhi      Cochin   

                       Route Dep_Time Arrival_Time Duration Total_Stops  Price  
0                  BLR → DEL    22:20        01:10   2h 50m    non-stop   3897  
1      CCU → IXR → BBI → BLR    05:50        13:15   7h 25m     2 stops   7662  
2      DEL → LKO → BOM → COK    09:25        04:25      19h     2 stops  13882  
3            CCU → NAG → BLR    18:05        23:30   5h 25m      1 stop   6218  
4            BLR → NAG → DEL    16:50        21:35   4h 45m      1 stop  13302  
...                      ...      ...          ...      ...         ...    ...  
10678              CCU → BLR    19:55        22:25   2h 30m    non-stop   4107  
10679              CCU → BLR    20:45        23:20   2h 35m    non-stop   4145  
10680              BLR → DEL    08:20        11:20       3h    non-stop   7229  
10681              BLR → DEL    11:30        14:10   2h 40m    non-stop  12648  
10682  DEL → GOI → BOM → COK    10:55        19:15   8h 20m     2 stops  11753  

[10683 rows x 10 columns]

### FEATURE ENGINEERING

Converting the data usign feature engineering

In [29]:
#Extracting day, month, and year from date of journey in flight price.

df1['Date']=df1['Date_of_Journey'].str.split('/').str[0]
df1['Month']=df1['Date_of_Journey'].str.split('/').str[1]
df1['Year']=df1['Date_of_Journey'].str.split('/').str[2]
df1.head()

Airline Date_of_Journey     Source Destination                  Route  \
0       IndiGo      24/03/2019   Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019    Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019  New Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019    Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019   Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time Arrival_Time Duration Total_Stops  Price Date Month  Year  
0    22:20        01:10   2h 50m    non-stop   3897   24    03  2019  
1    05:50        13:15   7h 25m     2 stops   7662    1    05  2019  
2    09:25        04:25      19h     2 stops  13882    9    06  2019  
3    18:05        23:30   5h 25m      1 stop   6218   12    05  2019  
4    16:50        21:35   4h 45m      1 stop  13302   01    03  2019

In [30]:
# Removing the "Date_of_Journey" column from flight price data

df1.drop('Date_of_Journey',axis=1,inplace=True)
df1.head()

Airline     Source Destination                  Route Dep_Time  \
0       IndiGo   Banglore   New Delhi              BLR → DEL    22:20   
1    Air India    Kolkata    Banglore  CCU → IXR → BBI → BLR    05:50   
2  Jet Airways  New Delhi      Cochin  DEL → LKO → BOM → COK    09:25   
3       IndiGo    Kolkata    Banglore        CCU → NAG → BLR    18:05   
4       IndiGo   Banglore   New Delhi        BLR → NAG → DEL    16:50   

  Arrival_Time Duration Total_Stops  Price Date Month  Year  
0        01:10   2h 50m    non-stop   3897   24    03  2019  
1        13:15   7h 25m     2 stops   7662    1    05  2019  
2        04:25      19h     2 stops  13882    9    06  2019  
3        23:30   5h 25m      1 stop   6218   12    05  2019  
4        21:35   4h 45m      1 stop  13302   01    03  2019

In [31]:
#Converting day, month, and year columns from object to integer data type in flight price data.

df1['Date']=df1['Date'].astype(int)
df1['Month']=df1['Month'].astype(int)
df1['Year']=df1['Year'].astype(int)
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Airline       10683 non-null  object
 1   Source        10683 non-null  object
 2   Destination   10683 non-null  object
 3   Route         10683 non-null  object
 4   Dep_Time      10683 non-null  object
 5   Arrival_Time  10683 non-null  object
 6   Duration      10683 non-null  object
 7   Total_Stops   10683 non-null  object
 8   Price         10683 non-null  int64 
 9   Date          10683 non-null  int32 
 10  Month         10683 non-null  int32 
 11  Year          10683 non-null  int32 
dtypes: int32(3), int64(1), object(8)
memory usage: 876.5+ KB


In [32]:
#Extracting hour and minute values from the 'Dep_Time' and creating new columns called 'Departure_hour' and 'Departure_min'

df1['Departure_hour']=df1.Dep_Time.str.split(':').str[0]
df1['Departure_min']=df1.Dep_Time.str.split(':').str[1]
df1.head()

Airline     Source Destination                  Route Dep_Time  \
0       IndiGo   Banglore   New Delhi              BLR → DEL    22:20   
1    Air India    Kolkata    Banglore  CCU → IXR → BBI → BLR    05:50   
2  Jet Airways  New Delhi      Cochin  DEL → LKO → BOM → COK    09:25   
3       IndiGo    Kolkata    Banglore        CCU → NAG → BLR    18:05   
4       IndiGo   Banglore   New Delhi        BLR → NAG → DEL    16:50   

  Arrival_Time Duration Total_Stops  Price  Date  Month  Year Departure_hour  \
0        01:10   2h 50m    non-stop   3897    24      3  2019             22   
1        13:15   7h 25m     2 stops   7662     1      5  2019             05   
2        04:25      19h     2 stops  13882     9      6  2019             09   
3        23:30   5h 25m      1 stop   6218    12      5  2019             18   
4        21:35   4h 45m      1 stop  13302     1      3  2019             16   

  Departure_min  
0            20  
1            50  
2            25  
3            05  
4            50

In [33]:
# Removing the "Dep_Time" column from flight price data

df1.drop('Dep_Time',axis=1,inplace=True)
df1.head()

Airline     Source Destination                  Route Arrival_Time  \
0       IndiGo   Banglore   New Delhi              BLR → DEL        01:10   
1    Air India    Kolkata    Banglore  CCU → IXR → BBI → BLR        13:15   
2  Jet Airways  New Delhi      Cochin  DEL → LKO → BOM → COK        04:25   
3       IndiGo    Kolkata    Banglore        CCU → NAG → BLR        23:30   
4       IndiGo   Banglore   New Delhi        BLR → NAG → DEL        21:35   

  Duration Total_Stops  Price  Date  Month  Year Departure_hour Departure_min  
0   2h 50m    non-stop   3897    24      3  2019             22            20  
1   7h 25m     2 stops   7662     1      5  2019             05            50  
2      19h     2 stops  13882     9      6  2019             09            25  
3   5h 25m      1 stop   6218    12      5  2019             18            05  
4   4h 45m      1 stop  13302     1      3  2019             16            50

In [34]:
#Extracting hour and minute values from the 'Arrival_Time' and creating new columns called 'Arrival_hour' and 'Arrival_min

df1['Arrival_hour']=df1.Arrival_Time.str.split(':').str[0]
df1['Arrival_min']=df1.Arrival_Time.str.split(':').str[1]
df1.head()

Airline     Source Destination                  Route Arrival_Time  \
0       IndiGo   Banglore   New Delhi              BLR → DEL        01:10   
1    Air India    Kolkata    Banglore  CCU → IXR → BBI → BLR        13:15   
2  Jet Airways  New Delhi      Cochin  DEL → LKO → BOM → COK        04:25   
3       IndiGo    Kolkata    Banglore        CCU → NAG → BLR        23:30   
4       IndiGo   Banglore   New Delhi        BLR → NAG → DEL        21:35   

  Duration Total_Stops  Price  Date  Month  Year Departure_hour Departure_min  \
0   2h 50m    non-stop   3897    24      3  2019             22            20   
1   7h 25m     2 stops   7662     1      5  2019             05            50   
2      19h     2 stops  13882     9      6  2019             09            25   
3   5h 25m      1 stop   6218    12      5  2019             18            05   
4   4h 45m      1 stop  13302     1      3  2019             16            50   

  Arrival_hour Arrival_min  
0           01          10  
1           13          15  
2           04          25  
3           23          30  
4           21          35

In [35]:
# Removing the "Arrival_Time" column from flight price data

df1.drop('Arrival_Time',axis=1,inplace=True)
df1.head()

Airline     Source Destination                  Route Duration  \
0       IndiGo   Banglore   New Delhi              BLR → DEL   2h 50m   
1    Air India    Kolkata    Banglore  CCU → IXR → BBI → BLR   7h 25m   
2  Jet Airways  New Delhi      Cochin  DEL → LKO → BOM → COK      19h   
3       IndiGo    Kolkata    Banglore        CCU → NAG → BLR   5h 25m   
4       IndiGo   Banglore   New Delhi        BLR → NAG → DEL   4h 45m   

  Total_Stops  Price  Date  Month  Year Departure_hour Departure_min  \
0    non-stop   3897    24      3  2019             22            20   
1     2 stops   7662     1      5  2019             05            50   
2     2 stops  13882     9      6  2019             09            25   
3      1 stop   6218    12      5  2019             18            05   
4      1 stop  13302     1      3  2019             16            50   

  Arrival_hour Arrival_min  
0           01          10  
1           13          15  
2           04          25  
3           23          30  
4           21          35

In [36]:
#Converting hour and minute columns for Departure and Arrival columns from object to integer data type in flight price data

df1['Departure_hour']=df1.Departure_hour.astype(int)
df1['Departure_min']=df1.Departure_min.astype(int)
df1['Arrival_hour']=df1.Arrival_hour.astype(int)
df1['Arrival_min']=df1.Arrival_min.astype(int)
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Airline         10683 non-null  object
 1   Source          10683 non-null  object
 2   Destination     10683 non-null  object
 3   Route           10683 non-null  object
 4   Duration        10683 non-null  object
 5   Total_Stops     10683 non-null  object
 6   Price           10683 non-null  int64 
 7   Date            10683 non-null  int32 
 8   Month           10683 non-null  int32 
 9   Year            10683 non-null  int32 
 10  Departure_hour  10683 non-null  int32 
 11  Departure_min   10683 non-null  int32 
 12  Arrival_hour    10683 non-null  int32 
 13  Arrival_min     10683 non-null  int32 
dtypes: int32(7), int64(1), object(6)
memory usage: 876.5+ KB


In [37]:
#Extracting hour values from the 'Duration' column in flight price data and creating a new column called 'Duration_hour'

df1['Duration_hour']=df1['Duration'].str.split(' ').str[0].str.split('h').str[0]
df1['Duration_hour']

0         2
1         7
2        19
3         5
4         4
         ..
10678     2
10679     2
10680     3
10681     2
10682     8
Name: Duration_hour, Length: 10683, dtype: object

In [38]:
 # Extracting minute values from the 'Duration' column in flight price data and creating a new column called 'Duration_min'
df1['Duration_min']=df1['Duration'].str.split(' ').str[1].str.split('m').str[0]
df1['Duration_min']

0         50
1         25
2        NaN
3         25
4         45
        ... 
10678     30
10679     35
10680    NaN
10681     40
10682     20
Name: Duration_min, Length: 10683, dtype: object

In [39]:
''' Some of the duration values in the original 'Duration' column were in hours only and did not have any minutes specified, 
which resulted in null values when the code tried to extract minute values from those rows.'''

df1.isnull().sum()

Airline              0
Source               0
Destination          0
Route                0
Duration             0
Total_Stops          0
Price                0
Date                 0
Month                0
Year                 0
Departure_hour       0
Departure_min        0
Arrival_hour         0
Arrival_min          0
Duration_hour        0
Duration_min      1031
dtype: int64

In [40]:
# Filling in null values in the 'Duration_min' column with zeros.

df1['Duration_min'].fillna(0,inplace=True)

In [41]:
df1.isnull().sum()

Airline           0
Source            0
Destination       0
Route             0
Duration          0
Total_Stops       0
Price             0
Date              0
Month             0
Year              0
Departure_hour    0
Departure_min     0
Arrival_hour      0
Arrival_min       0
Duration_hour     0
Duration_min      0
dtype: int64

In [42]:
# Removing the original 'Duration' column from the flight price data

df1.drop('Duration',axis=1,inplace=True)

In [43]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Airline         10683 non-null  object
 1   Source          10683 non-null  object
 2   Destination     10683 non-null  object
 3   Route           10683 non-null  object
 4   Total_Stops     10683 non-null  object
 5   Price           10683 non-null  int64 
 6   Date            10683 non-null  int32 
 7   Month           10683 non-null  int32 
 8   Year            10683 non-null  int32 
 9   Departure_hour  10683 non-null  int32 
 10  Departure_min   10683 non-null  int32 
 11  Arrival_hour    10683 non-null  int32 
 12  Arrival_min     10683 non-null  int32 
 13  Duration_hour   10683 non-null  object
 14  Duration_min    10683 non-null  object
dtypes: int32(7), int64(1), object(7)
memory usage: 959.9+ KB


In [44]:
#Converting the 'Duration_hour' and 'Duration_min' columns from object to integer data type in flight price data

df1['Duration_hour']=df1['Duration_hour'].astype(int)
df1['Duration_min']=df1['Duration_min'].astype(int)


In [45]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Airline         10683 non-null  object
 1   Source          10683 non-null  object
 2   Destination     10683 non-null  object
 3   Route           10683 non-null  object
 4   Total_Stops     10683 non-null  object
 5   Price           10683 non-null  int64 
 6   Date            10683 non-null  int32 
 7   Month           10683 non-null  int32 
 8   Year            10683 non-null  int32 
 9   Departure_hour  10683 non-null  int32 
 10  Departure_min   10683 non-null  int32 
 11  Arrival_hour    10683 non-null  int32 
 12  Arrival_min     10683 non-null  int32 
 13  Duration_hour   10683 non-null  int32 
 14  Duration_min    10683 non-null  int32 
dtypes: int32(9), int64(1), object(5)
memory usage: 876.5+ KB


In [46]:
df1.Total_Stops.unique()

array(['non-stop', '2 stops', '1 stop', '3 stops', '4 stops'],
      dtype=object)

In [47]:
# Mapping  the string values in the 'Total_Stops' column to their corresponding numeric values.

df1['Total_Stops']=df1['Total_Stops'].map({'non-stop':0, '1 stop':1, '2 stops':2, '3 stops':3, '4 stops':4})
df1.head()

Airline     Source Destination                  Route  Total_Stops  \
0       IndiGo   Banglore   New Delhi              BLR → DEL            0   
1    Air India    Kolkata    Banglore  CCU → IXR → BBI → BLR            2   
2  Jet Airways  New Delhi      Cochin  DEL → LKO → BOM → COK            2   
3       IndiGo    Kolkata    Banglore        CCU → NAG → BLR            1   
4       IndiGo   Banglore   New Delhi        BLR → NAG → DEL            1   

   Price  Date  Month  Year  Departure_hour  Departure_min  Arrival_hour  \
0   3897    24      3  2019              22             20             1   
1   7662     1      5  2019               5             50            13   
2  13882     9      6  2019               9             25             4   
3   6218    12      5  2019              18              5            23   
4  13302     1      3  2019              16             50            21   

   Arrival_min  Duration_hour  Duration_min  
0           10              2            50  
1           15              7            25  
2           25             19             0  
3           30              5            25  
4           35              4            45

In [48]:
df1['Total_Stops']=df1['Total_Stops'].astype(int)
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Airline         10683 non-null  object
 1   Source          10683 non-null  object
 2   Destination     10683 non-null  object
 3   Route           10683 non-null  object
 4   Total_Stops     10683 non-null  int32 
 5   Price           10683 non-null  int64 
 6   Date            10683 non-null  int32 
 7   Month           10683 non-null  int32 
 8   Year            10683 non-null  int32 
 9   Departure_hour  10683 non-null  int32 
 10  Departure_min   10683 non-null  int32 
 11  Arrival_hour    10683 non-null  int32 
 12  Arrival_min     10683 non-null  int32 
 13  Duration_hour   10683 non-null  int32 
 14  Duration_min    10683 non-null  int32 
dtypes: int32(10), int64(1), object(4)
memory usage: 834.7+ KB


In [49]:
# Total_stops and Routes have same functioning so dropping the route column.

df1.drop('Route',axis=1,inplace=True)

In [50]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Airline         10683 non-null  object
 1   Source          10683 non-null  object
 2   Destination     10683 non-null  object
 3   Total_Stops     10683 non-null  int32 
 4   Price           10683 non-null  int64 
 5   Date            10683 non-null  int32 
 6   Month           10683 non-null  int32 
 7   Year            10683 non-null  int32 
 8   Departure_hour  10683 non-null  int32 
 9   Departure_min   10683 non-null  int32 
 10  Arrival_hour    10683 non-null  int32 
 11  Arrival_min     10683 non-null  int32 
 12  Duration_hour   10683 non-null  int32 
 13  Duration_min    10683 non-null  int32 
dtypes: int32(10), int64(1), object(3)
memory usage: 751.3+ KB


In [51]:
'''One-hot encoding is a technique used to transform categorical data into a format that can be more easily 
interpreted by machine learning algorithms. The OneHotEncoder class creates a binary matrix for each unique category in a 
categorical variable, where each row represents an observation and each column represents a category'''

from sklearn.preprocessing import OneHotEncoder
encoder=OneHotEncoder()

In [52]:
#Array containing the one-hot encoded values for the specified columns in the input DataFrame

encoder.fit_transform(df1[['Airline','Source','Destination']]).toarray()

array([[0., 0., 0., ..., 0., 0., 1.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 1.],
       [0., 0., 0., ..., 0., 0., 1.],
       [0., 1., 0., ..., 0., 0., 0.]])

In [53]:
ohe=pd.DataFrame(encoder.fit_transform(df1[['Airline','Source','Destination']]).toarray(),columns=encoder.get_feature_names_out())

In [54]:
df1.drop(['Airline','Source','Destination'],axis=1,inplace=True)

In [55]:
df1.head()

Total_Stops  Price  Date  Month  Year  Departure_hour  Departure_min  \
0            0   3897    24      3  2019              22             20   
1            2   7662     1      5  2019               5             50   
2            2  13882     9      6  2019               9             25   
3            1   6218    12      5  2019              18              5   
4            1  13302     1      3  2019              16             50   

   Arrival_hour  Arrival_min  Duration_hour  Duration_min  
0             1           10              2            50  
1            13           15              7            25  
2             4           25             19             0  
3            23           30              5            25  
4            21           35              4            45

In [56]:
# Combining the categorical converted columns with rest of df.

df1=pd.concat([df1,ohe],axis=1)

In [57]:
df1

Total_Stops  Price  Date  Month  Year  Departure_hour  Departure_min  \
0                0   3897    24      3  2019              22             20   
1                2   7662     1      5  2019               5             50   
2                2  13882     9      6  2019               9             25   
3                1   6218    12      5  2019              18              5   
4                1  13302     1      3  2019              16             50   
...            ...    ...   ...    ...   ...             ...            ...   
10678            0   4107     9      4  2019              19             55   
10679            0   4145    27      4  2019              20             45   
10680            0   7229    27      4  2019               8             20   
10681            0  12648     1      3  2019              11             30   
10682            2  11753     9      5  2019              10             55   

       Arrival_hour  Arrival_min  Duration_hour  ...  Source_Banglore  \
0                 1           10              2  ...              1.0   
1                13           15              7  ...              0.0   
2                 4           25             19  ...              0.0   
3                23           30              5  ...              0.0   
4                21           35              4  ...              1.0   
...             ...          ...            ...  ...              ...   
10678            22           25              2  ...              0.0   
10679            23           20              2  ...              0.0   
10680            11           20              3  ...              1.0   
10681            14           10              2  ...              1.0   
10682            19           15              8  ...              0.0   

       Source_Chennai  Source_Kolkata  Source_Mumbai  Source_New Delhi  \
0                 0.0             0.0            0.0               0.0   
1                 0.0             1.0            0.0               0.0   
2                 0.0             0.0            0.0               1.0   
3                 0.0             1.0            0.0               0.0   
4                 0.0             0.0            0.0               0.0   
...               ...             ...            ...               ...   
10678             0.0             1.0            0.0               0.0   
10679             0.0             1.0            0.0               0.0   
10680             0.0             0.0            0.0               0.0   
10681             0.0             0.0            0.0               0.0   
10682             0.0             0.0            0.0               1.0   

       Destination_Banglore  Destination_Cochin  Destination_Hyderabad  \
0                       0.0                 0.0                    0.0   
1                       1.0                 0.0                    0.0   
2                       0.0                 1.0                    0.0   
3                       1.0                 0.0                    0.0   
4                       0.0                 0.0                    0.0   
...                     ...                 ...                    ...   
10678                   1.0                 0.0                    0.0   
10679                   1.0                 0.0                    0.0   
10680                   0.0                 0.0                    0.0   
10681                   0.0                 0.0                    0.0   
10682                   0.0                 1.0                    0.0   

       Destination_Kolkata  Destination_New Delhi  
0                      0.0                    1.0  
1                      0.0                    0.0  
2                      0.0                    0.0  
3                      0.0                    0.0  
4                      0.0                    1.0  
...                    ...                    ...  
10678                  0.0             

In [58]:
df1.to_csv('Flight_price_cleaned_EDA.csv',index=False)